### ISSUE TITLE
200 status code is returned even when server times out

### ISSUE THREAD

**Issue by dkoo761** [prerender/prerender#515]
Hey there,

My use case is that I'm using prerender to quickly generate a preview of a website as part of a server-side rendering (SSR) solution.

I'm running the latest prerender server release and noticed some strange output in the logs:

```
2018-05-24T01:54:03.785Z page timed out http://localhost:3000
....
2018-05-24T01:54:04.391Z got 200 in 5772ms for http://localhost:3000
```

The response that I get back from the HTTP request to my local prerender server also has `statusCode: 200`.
If the prerender server timed out before the page was finished loading, shouldn't the response be something other than 200 since the response body will probably not contain the correct/complete HTML? Basically I need a way to know if the page didn't render correctly so that I can retry the prerender rather than displaying the wrong HTML to the user.

**Comment by thoop** [comment]
First, be careful about returning a prerendered page to users since our prerender server doesn't do any serializing/deserializing of javascript state from the SSR to the user's browser. You might run into some weird things like full page refreshes.

That being said, the timeout defaults to 20 seconds and is only meant to prevent a page from loading forever. Since the Prerender server counts the number of requests in flight and waits until a short time after the number of requests in flight reaches zero to save the HTML, when the timeout is hit that usually means there were some requests in flight the entire time preventing us from saving the page. This is usually seen as pinging an API url over and over or is sometimes seen if `window.prerenderReady` is set to `false` and never set to `true`.

If you need a different status code on the response, you can utilize our `prerender-status-code` meta tag. You might do something like initialize the page with `<meta name="prerender-status-code" content="503">` in the head of your page and then after you know your page has rendered correctly you would set `<meta name="prerender-status-code" content="200">`. That way if the page was ever rendered before you knew it was finished, you would see a 503 and could retry.

Let me know if you have any questions about that.

**Comment by dkoo761** [comment]
Hey Todd, thanks for the info. Yes, I did have a couple window.prerenderReady's missing earlier and have since fixed those. Sometimes pages still time out though (even with a 60 second value) and since we're serving these to end users it's important to know when those timeouts happen.

I didn't know about the `prerender-status-code`, cool! I ended up just making a small change to the prerender package to return a 504 after a timeout which works perfectly for our needs but I might change it to the meta tag instead.
